### PR TITLE
Add possibility to specify multiple disks in 'virt' module

### PR DESCRIPTION
**Pull request by dis-xcom**
Now disk profiles can contain multiple disks:

- backward compatible: if 'image' argument is set for virt.init()
  *and* 'image' attribute is missing for the first disk, then the
  first disk will use the 'image' from the method arguments

- multiple disks can be specified for a disk profile, see pillar
  example below

- if 'size' attribute set for qcow2-based image, then resulting
  image will be resized to the 'size' capacity (author: Ondrej Smola)

- empty disks allowed ('size' attribute is mandatory in this case)

Example disk profile 'three_disks':

    virt:
      disk:
        three_disks:
          - system:
              size: 4096
              image: http://path/to/ubuntu.qcow2
          - repository_snapshot:
              size: 8192
              image: /var/lib/libvirt/images/snapshot.qcow2
          - cinder-volume:
              size: 16384

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

File: salt/modules/virt.py [hunk 1]
```diff
@@ -673,6 +634,117 @@ def init(name,
...
+                    disk_file_name
+                )
+                log.debug('Image destination will be {0}'.format(img_dest))
+                img_dir = os.path.dirname(img_dest)
+                log.debug('Image destination directory is {0}'
+                          .format(img_dir))
+                try:
+                    os.makedirs(img_dir)
+                except OSError:
+                    pass
```
Comment (cachedout) [inline salt/modules/virt.py]: Are we sure we want to just ignore this?

File: salt/modules/virt.py [hunk 2]
```diff
@@ -673,6 +634,117 @@ def init(name,
                     )
                     define_vol_xml_str(xml)
 
+            elif hypervisor in ['qemu', 'kvm']:
+
```
Comment (tgoodlet) [inline salt/modules/virt.py]: I'd say stick the ` for disk in diskp:` somewhere here; only loop when using the `libvirt` hypervisor.
Sticking lines 639-740 in a new function might make this a lot cleaner as well.

File: salt/modules/virt.py [hunk 3]
```diff
@@ -574,96 +592,39 @@ def init(name,
...
-                qcow2 = imageinfo['file format'] == 'qcow2'
-
-            try:
-                if enable_qcow and qcow2:
-                    log.info('Cloning qcow2 image {} using copy on write'
-                              .format(sfn))
-                    __salt__['cmd.run'](
-                        'qemu-img create -f qcow2 -o backing_file={} {}'
-                        .format(sfn, img_dest).split())
+        if not diskp[0][disk_name].get('image', None):
```
Comment (tgoodlet) [inline salt/modules/virt.py]: So if a user passes an `image` arg we only override it if the first entry in the profile **isn't defined**?
That seems like incorrect precedence; the arg should always take priority.

File: salt/modules/virt.py [hunk 4]
```diff
@@ -574,96 +592,39 @@ def init(name,
...
+            log.debug('{0} will be used as default image'.format(image))
+            diskp[0][disk_name]['image'] = image
+
+    # Create multiple disks, empty or from specified images.
+    for disk in diskp:
```
Comment (tgoodlet) [inline salt/modules/virt.py]: Not following why this loop was moved here?
Do you need it if `hypervisor not in ['qemu', 'kvm']`?

I'm wondering if maybe it's worth moving all the 639-740 bit into a new function anyway; the indentation is getting a bit nested/convoluted.

**Review by goodboy — CHANGES_REQUESTED** [review]
Some other notes:
- The legacy esxi/vmware stuff is probably never going to be added [(especially since `salt-cloud` already supports it)](https://docs.saltstack.com/en/latest/topics/cloud/vmware.html).
- the [libvirt backend for `salt-cloud`](https://github.com/saltstack/salt/pull/36709) might be a better place to stick this.
I know that I would have probably put my changes against that system if I had known about it at the time.

File: salt/modules/virt.py (hunk 1 above)
Comment (dis-xcom) [inline salt/modules/virt.py]: This is the original code (just moved, see L620-623 in original file). I didn't analyze why this code was added early, so leave it as-is for now.

File: salt/modules/virt.py (hunk 3 above)
Comment (dis-xcom) [inline salt/modules/virt.py]: Default disk profile doesn't have 'image' attribute, so will always use 'image' from the module arguments.

But if a disk profile uses different images for different disks, there is no determined way to specify which images should be replaced with the image from argument:
disk_a:
    image: aaaa.qcow2
disk_b:
   image: bbbb.qcow2
disk_c:
   image: cccc.qcow2

If you think that the 'image' module argument should be for the first disk only, then it should be reflected at least in the docstring

File: salt/modules/virt.py (hunk 4 above)
Comment (dis-xcom) [inline salt/modules/virt.py]: Yes, we need it, see L665 in the original file (this code is reached if we have multiple-disks profile without specified images for non- qemu-kvm hypervisors).

There are two branches that can be nested in different orders:
- check for hypervisor and then iterate over disks for each hypervisor separatelly
- iterate over disks and check for hypervisor for creating the disks

The first option requires two different iteration loops in different places, the second option have a single iteration over disks in the disk profile. I think the second option will be more readable and easily to support.

As for moving the image creation code into a separated method - good idea, thanks!

File: salt/modules/virt.py (hunk 2 above)
Comment (dis-xcom) [inline salt/modules/virt.py]: As I said - we should iterate 'diskp' for any type of hypervisors, the difference is: for qemu-kvm we create image, for other hypervisors we just describe the volume for libvirt (define_vol_xml_str() )

File: salt/modules/virt.py (hunk 1 above)
Comment (dis-xcom) [inline salt/modules/virt.py]: Here is that commit: https://github.com/saltstack/salt/commit/6413c11f2946c20676bae69f63d539e8ba44ebea#diff-b651f4b5ad5aad0bb9fd22bacd93caa1R593

**Comment by cachedout** [comment]
@tgoodlet How are you feeling about this one after the latest changes?

File: salt/modules/virt.py [hunk 5]
```diff
@@ -580,90 +700,34 @@ def init(name,
...
+                          diskp[0][disk_name].get('image', None)))
+        diskp[0][disk_name]['image'] = image
+
+    # Create multiple disks, empty or from specified images.
+    for disk in diskp:
+        log.debug("Creating disk for VM [ {0} ]: {1}".format(name, disk))
+
+        for disk_name, args in six.iteritems(disk):
+
+            if hypervisor in ['esxi', 'vmware']:
```
Comment (tgoodlet) [inline salt/modules/virt.py]: @cachedout Can't we just drop this?

I mean is anyone actually going to add it given that `salt-cloud` is turning out to be the new system for all this?

File: salt/modules/virt.py [hunk 6]
```diff
@@ -353,6 +353,108 @@ def _qemu_image_info(path):
...
+                    'qemu-img create -f qcow2 -o backing_file={0} {1}'
+                    .format(sfn, img_dest).split())
+            else:
+                log.debug('Copying {0} to {1}'.format(sfn, img_dest))
+                salt.utils.files.copyfile(sfn, img_dest)
+
+            mask = os.umask(0)
+            os.umask(mask)
+
+            if disk_size and qcow2:
```
Comment (tgoodlet) [inline salt/modules/virt.py]: It seems odd to be the function needs to be called with something `False`y in order to avoid resizing? Maybe just set `disk_size=None`?

File: salt/modules/virt.py [hunk 7]
```diff
@@ -673,6 +737,49 @@ def init(name,
...
+                disk_file_name = '{0}.{1}'.format(disk_name, disk_type)
+
+                if not disk_size and not disk_image:
+                    raise CommandExecutionError(
+                        'Unable to create new disk {0}, please specify'
+                        ' at least <size> and/or <image> argument'
+                        .format(disk_file_name)
+                    )
+
+                img_dest = _qemu_image_create(
```
Comment (tgoodlet) [inline salt/modules/virt.py]: Yeah this is looking a lot easier to grok :)

File: salt/modules/virt.py (hunk 6 above)
Comment (dis-xcom) [inline salt/modules/virt.py]: Just don't set 'size' attribute in the disk profile to avoid resizing, and it will be initialized with None in L744.

For standalone use of this method your comment makes sense, thanks!

File: salt/modules/virt.py (hunk 5 above)
Comment (dis-xcom) [inline salt/modules/virt.py]: I would prefer to make such changes in a different patch, because it could introduce regression that is not connected to the feature I'm trying to add :)
